File: conduit/tests/python/test_ebcdic_coverage.py
"""Roundtrip encode/decode tests for EBCDIC string generated Python module.

Covers:
- ebcdic_test: EbcdicMsg with id (uint8), label (EBCDIC string, 8 bytes,
  space-padded, right-trimmed), trailer (uint16)
"""
import pytest


# ============================================================================
# ebcdic_test: EbcdicMsg roundtrip tests
# ============================================================================


class TestEbcdicMsgBasicRoundtrip:
    """Basic encode/decode roundtrip for EBCDIC string fields."""

    def test_simple_label(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0x42
        msg.label = "HELLO"
        msg.trailer = 0x1234

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.id == 0x42
        assert msg2.label.rstrip() == "HELLO"
        assert msg2.trailer == 0x1234

    def test_full_length_label(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0x01
        msg.label = "ABCDEFGH"  # exactly 8 chars
        msg.trailer = 0xFFFF

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.label == "ABCDEFGH"
        assert msg2.trailer == 0xFFFF

    def test_single_char_label(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0x02
        msg.label = "X"
        msg.trailer = 100

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.label.rstrip() == "X"
        assert msg2.trailer == 100


class TestEbcdicMsgEmptyLabel:
    """Empty label should be space-padded and trimmed back to empty."""

    def test_empty_label(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0x03
        msg.label = ""
        msg.trailer = 0

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.label.rstrip() == ""
        assert msg2.id == 0x03
        assert msg2.trailer == 0


class TestEbcdicMsgBoundaryValues:
    """Boundary values for surrounding numeric fields."""

    def test_id_max(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0xFF
        msg.label = "TEST"
        msg.trailer = 0

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.id == 0xFF

    def test_id_zero(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0
        msg.label = "TEST"
        msg.trailer = 0xFFFF

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.id == 0
        assert msg2.trailer == 0xFFFF

    def test_trailer_max(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 1
        msg.label = "MAX"
        msg.trailer = 0xFFFF

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.trailer == 0xFFFF


class TestEbcdicMsgWireFormat:
    """Verify wire format size: 1 (id) + 8 (label) + 2 (trailer) = 11 bytes."""

    def test_wire_size(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0x10
        msg.label = "SIZE"
        msg.trailer = 0x20

        data = msg.encode_bytes()
        assert len(data) == 11  # 1 + 8 + 2


class TestEbcdicMsgDoubleEncode:
    """Encode -> decode -> encode should produce identical bytes."""

    def test_double_encode_stability(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0xAB
        msg.label = "STABLE"
        msg.trailer = 0xCDEF

        data1 = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2

    def test_double_encode_empty_label(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0
        msg.label = ""
        msg.trailer = 0

        data1 = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2

    def test_double_encode_full_label(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 0xFF
        msg.label = "ABCDEFGH"
        msg.trailer = 0xFFFF

        data1 = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data1)
        data2 = msg2.encode_bytes()
        assert data1 == data2


class TestEbcdicMsgVariousStrings:
    """Test various string content through EBCDIC encoding."""

    def test_numeric_string(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 1
        msg.label = "12345678"
        msg.trailer = 0

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.label == "12345678"

    def test_uppercase_alpha(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 2
        msg.label = "ABCDEFGH"
        msg.trailer = 0

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.label == "ABCDEFGH"

    def test_mixed_content(self):
        from ebcdic_test import EbcdicMsg

        msg = EbcdicMsg()
        msg.id = 3
        msg.label = "A1B2C3"
        msg.trailer = 0

        data = msg.encode_bytes()
        msg2 = EbcdicMsg.decode_bytes(data)
        assert msg2.label.rstrip() == "A1B2C3"

File: conduit/tests/java/TestEbcdicCoverage.java
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for EBCDIC string encoding roundtrip in generated Java codecs.
 */
public class TestEbcdicCoverage {

    // ========================================================================
    // ebcdic_strings: EbcdicMsg with EBCDIC-encoded label field
    // ========================================================================

    @Test
    @DisplayName("EbcdicMsg: roundtrip with empty string (space-padded)")
    void ebcdicEmptyString() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 0;
        msg.label = "";
        msg.trailer = 0;
        byte[] encoded = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(encoded);
        assertEquals(0, d.id);
        assertTrue(d.label.trim().isEmpty());
        assertEquals(0, d.trailer);
    }

    @Test
    @DisplayName("EbcdicMsg: single character roundtrip")
    void ebcdicSingleChar() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 1;
        msg.label = "X";
        msg.trailer = 0x0001;
        byte[] encoded = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(encoded);
        assertEquals("X", d.label.trim());
        assertEquals(0x0001, d.trailer);
    }

    @Test
    @DisplayName("EbcdicMsg: full-length 8-char string roundtrip")
    void ebcdicFullLength() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 2;
        msg.label = "ABCDEFGH";
        msg.trailer = 0xFFFF;
        byte[] encoded = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(encoded);
        assertEquals("ABCDEFGH", d.label.trim());
        assertEquals(0xFFFF, d.trailer);
    }

    @Test
    @DisplayName("EbcdicMsg: short string is right-padded with spaces")
    void ebcdicShortString() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 3;
        msg.label = "HELLO";
        msg.trailer = 0x1234;
        byte[] encoded = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(encoded);
        assertEquals("HELLO", d.label.trim());
        assertEquals(0x1234, d.trailer);
    }

    @Test
    @DisplayName("EbcdicMsg: numeric string content")
    void ebcdicNumericString() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 4;
        msg.label = "12345678";
        msg.trailer = 0;
        byte[] encoded = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(encoded);
        assertEquals("12345678", d.label.trim());
    }

    @Test
    @DisplayName("EbcdicMsg: mixed alphanumeric content")
    void ebcdicMixedContent() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 5;
        msg.label = "AB12CD34";
        msg.trailer = 0x5678;
        byte[] encoded = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(encoded);
        assertEquals("AB12CD34", d.label.trim());
    }

    @Test
    @DisplayName("EbcdicMsg: wire format is 11 bytes (1 + 8 + 2)")
    void ebcdicWireSize() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 0;
        msg.label = "";
        msg.trailer = 0;
        byte[] encoded = msg.encodeBytes();
        assertEquals(11, encoded.length);
    }

    @Test
    @DisplayName("EbcdicMsg: wire size unchanged regardless of string length")
    void ebcdicWireSizeConsistent() {
        ebcdic_strings.EbcdicMsg msg1 = new ebcdic_strings.EbcdicMsg();
        msg1.id = 0;
        msg1.label = "A";
        msg1.trailer = 0;

        ebcdic_strings.EbcdicMsg msg2 = new ebcdic_strings.EbcdicMsg();
        msg2.id = 0;
        msg2.label = "ABCDEFGH";
        msg2.trailer = 0;

        assertEquals(msg1.encodeBytes().length, msg2.encodeBytes().length);
    }

    @Test
    @DisplayName("EbcdicMsg: double-encode stability")
    void ebcdicDoubleEncode() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 0xFF;
        msg.label = "TEST";
        msg.trailer = 0xABCD;
        byte[] data1 = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(data1);
        byte[] data2 = d.encodeBytes();
        assertArrayEquals(data1, data2);
    }

    @Test
    @DisplayName("EbcdicMsg: double-encode stability with full-length string")
    void ebcdicDoubleEncodeFullLength() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 0x42;
        msg.label = "ROUNDTRP";
        msg.trailer = 0x9999;
        byte[] data1 = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(data1);
        byte[] data2 = d.encodeBytes();
        assertArrayEquals(data1, data2);
    }

    @Test
    @DisplayName("EbcdicMsg: max id boundary 0xFF")
    void ebcdicMaxId() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 0xFF;
        msg.label = "DATA";
        msg.trailer = 0;
        byte[] encoded = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(encoded);
        assertEquals(0xFF, d.id);
    }

    @Test
    @DisplayName("EbcdicMsg: zero id and zero trailer")
    void ebcdicZeroFields() {
        ebcdic_strings.EbcdicMsg msg = new ebcdic_strings.EbcdicMsg();
        msg.id = 0;
        msg.label = "ZERO";
        msg.trailer = 0;
        byte[] encoded = msg.encodeBytes();
        ebcdic_strings.EbcdicMsg d = ebcdic_strings.EbcdicMsg.decodeBytes(encoded);
        assertEquals(0, d.id);
        assertEquals("ZERO", d.label.trim());
        assertEquals(0, d.trailer);
    }
}
